File: tests/tests/ip_link/big_graph/big_graph.py
import sys
from collections import defaultdict

class CMakeGraphFlattener:
    def __init__(self):
        self.targets = {}
        self.global_stack = []
        self.rm_lists = defaultdict(list)
        
    def add_target(self, name, interface_link_libraries=None):
        """Simulate add_library/add_executable"""
        if interface_link_libraries is None:
            interface_link_libraries = []
        self.targets[name] = {
            'interface_link_libraries': interface_link_libraries,
            'is_target': True
        }
    
    def flatten_graph(self, node):
        """Main entry point - flatten the dependency graph starting from node"""
        node = self.alias_dereference(node)
        self.global_stack = []
        found_root = -1
        
        while found_root == -1:
            self.__flatten_graph_recursive(node)
            try:
                found_root = self.global_stack.index(node)
            except ValueError:
                found_root = -1
        
        # Clear RM lists
        for lib in self.global_stack:
            lib = self.alias_dereference(lib)
            self.rm_lists[lib] = []
            
        return self.global_stack
    
    def __flatten_graph_recursive(self, node):
        """Recursive helper for flattening"""
        node = self.alias_dereference(node)
        
        # If it's not a target (like -pthread)
        if node not in self.targets or not self.targets[node]['is_target']:
            return 1
        
        # If all dependencies have been processed
        if self.__all_vertices_removed(node):
            if node not in self.global_stack:
                self.global_stack.append(node)
            return 1
        
        # Get dependencies, filtering out special patterns
        deps = self.targets[node]['interface_link_libraries']
        deps = [d for d in deps if not d.endswith("::@")]
        
        for lib in deps:
            lib_added = self.__flatten_graph_recursive(lib)
            if lib_added == 1:
                self.__append_rm_list_unique(node, lib)
        
        return 0
    
    def __append_rm_list_unique(self, node, rm_el):
        """Add to removal list if not already present"""
        node = self.alias_dereference(node)
        if rm_el not in self.rm_lists[node]:
            self.rm_lists[node].append(rm_el)
    
    def __all_vertices_removed(self, node):
        """Check if all dependencies have been processed"""
        if node not in self.targets:
            raise ValueError(f"Node is not defined {node}")
        
        rm_list = self.rm_lists.get(node, [])
        link_libs = self.targets[node]['interface_link_libraries']
        link_libs = [lib for lib in link_libs if not lib.endswith("::@")]
        link_libs = list(dict.fromkeys(link_libs))  # Remove duplicates
        
        return self.compare_lists(rm_list, link_libs)
    
    def compare_lists(self, l1, l2):
        """Compare if two lists contain the same elements"""
        if len(l1) != len(l2):
            return False
        return all(item in l2 for item in l1)
    
    def alias_dereference(self, node):
        """Simulate CMake's alias dereferencing"""
        # In a real implementation, this would resolve CMake aliases
        return node


def test_big_graph():
    """Test with a large graph similar to the CMake test case"""
    flattener = CMakeGraphFlattener()
    total_ips = 100000
    branching_factor = 2
    
    # Add root IP
    root_ip = "ip0"
    flattener.add_target(root_ip)
    ip_index = 1
    
    queue = [root_ip]
    
    # Build the graph
    while ip_index < total_ips:
        next_queue = []
        
        for parent_ip in queue:
            if ip_index >= total_ips:
                break
            
            children = []
            for _ in range(branching_factor):
                if ip_index >= total_ips:
                    break
                
                child_ip = f"ip{ip_index}"
                flattener.add_target(child_ip)
                children.append(child_ip)
                next_queue.append(child_ip)
                ip_index += 1
            
            if children:
                # Add dependencies
                flattener.targets[parent_ip]['interface_link_libraries'].extend(children)
        
        queue = next_queue
    
    # Flatten the graph
    flattened = flattener.flatten_graph(root_ip)
    
    # print(f"Flattened graph has {len(flattened)} nodes")
    # print("First 5:", flattened[:5])
    # print("Last 5:", flattened[-5:])
    #
    # # Verify we got all nodes
    # assert len(flattened) == total_ips
    # print("Test passed!")
    #
    # for i in flattened:
    #     print(i, end=";")
    #


if __name__ == "__main__":
    test_big_graph()

File: tests/tests/ip_link/big_graph/big_graph.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <string>
#include <algorithm>

class CMakeGraphFlattener {
public:
    struct Target {
        std::vector<std::string> interface_link_libraries;
        bool is_target = true;
    };

    void add_target(const std::string& name, 
                   const std::vector<std::string>& deps = {}) {
        targets_[name] = {deps, true};
    }

    std::vector<std::string> flatten_graph(const std::string& node) {
        std::string resolved_node = alias_dereference(node);
        global_stack_.clear();
        int found_root = -1;

        while (found_root == -1) {
            __flatten_graph_recursive(resolved_node);
            auto it = std::find(global_stack_.begin(), global_stack_.end(), resolved_node);
            found_root = (it != global_stack_.end()) ? 1 : -1;
        }

        // Clear RM lists
        for (const auto& lib : global_stack_) {
            rm_lists_[lib].clear();
        }

        return global_stack_;
    }

    int __flatten_graph_recursive(const std::string& node) {
        std::string resolved_node = alias_dereference(node);

        // If it's not a target (like -pthread)
        if (targets_.count(resolved_node) == 0 || !targets_[resolved_node].is_target) {
            return 1;
        }

        // If all dependencies have been processed
        if (__all_vertices_removed(resolved_node)) {
            if (std::find(global_stack_.begin(), global_stack_.end(), resolved_node) == global_stack_.end()) {
                global_stack_.push_back(resolved_node);
            }
            return 1;
        }

        // Get dependencies, filtering out special patterns
        auto deps = targets_[resolved_node].interface_link_libraries;
        deps.erase(std::remove_if(deps.begin(), deps.end(),
                   [](const std::string& s) { return s.find("::@") != std::string::npos; }),
                   deps.end());

        for (const auto& lib : deps) {
            int lib_added = __flatten_graph_recursive(lib);
            if (lib_added == 1) {
                __append_rm_list_unique(resolved_node, lib);
            }
        }

        return 0;
    }

    void __append_rm_list_unique(const std::string& node, const std::string& rm_el) {
        std::string resolved_node = alias_dereference(node);
        auto& rm_list = rm_lists_[resolved_node];
        if (std::find(rm_list.begin(), rm_list.end(), rm_el) == rm_list.end()) {
            rm_list.push_back(rm_el);
        }
    }

    bool __all_vertices_removed(const std::string& node) {
        if (targets_.count(node) == 0) {
            throw std::runtime_error("Node is not defined: " + node);
        }

        auto& rm_list = rm_lists_[node];
        auto link_libs = targets_[node].interface_link_libraries;

        // Filter and remove duplicates
        link_libs.erase(std::remove_if(link_libs.begin(), link_libs.end(),
                       [](const std::string& s) { return s.find("::@") != std::string::npos; }),
                       link_libs.end());
        std::sort(link_libs.begin(), link_libs.end());
        link_libs.erase(std::unique(link_libs.begin(), link_libs.end()), link_libs.end());

        return compare_lists(rm_list, link_libs);
    }

    bool compare_lists(const std::vector<std::string>& l1, 
                      const std::vector<std::string>& l2) {
        if (l1.size() != l2.size()) return false;

        auto l1_sorted = l1;
        auto l2_sorted = l2;
        std::sort(l1_sorted.begin(), l1_sorted.end());
        std::sort(l2_sorted.begin(), l2_sorted.end());

        return l1_sorted == l2_sorted;
    }

    std::string alias_dereference(const std::string& node) {
        // In a real implementation, this would resolve CMake aliases
        return node;
    }

    std::unordered_map<std::string, Target> targets_;
    std::unordered_map<std::string, std::vector<std::string>> rm_lists_;
    std::vector<std::string> global_stack_;
};

void test_big_graph() {
    CMakeGraphFlattener flattener;
    const int TOTAL_IPS = 100000;
    const int BRANCHING_FACTOR = 2;

    // Add root IP
    std::string root_ip = "ip0";
    flattener.add_target(root_ip);
    int ip_index = 1;

    std::vector<std::string> queue = {root_ip};

    // Build the graph
    while (ip_index < TOTAL_IPS) {
        std::vector<std::string> next_queue;

        for (const auto& parent_ip : queue) {
            if (ip_index >= TOTAL_IPS) break;

            std::vector<std::string> children;
            for (int i = 0; i < BRANCHING_FACTOR; i++) {
                if (ip_index >= TOTAL_IPS) break;

                std::string child_ip = "ip" + std::to_string(ip_index);
                flattener.add_target(child_ip);
                children.push_back(child_ip);
                next_queue.push_back(child_ip);
                ip_index++;
            }

            if (!children.empty()) {
                // Add dependencies
                auto& deps = flattener.targets_[parent_ip].interface_link_libraries;
                deps.insert(deps.end(), children.begin(), children.end());
            }
        }

        queue = next_queue;
    }

    // Flatten the graph
    auto flattened = flattener.flatten_graph(root_ip);

    std::cout << "Flattened graph has " << flattened.size() << " nodes\n";
    std::cout << "First 5: ";
    for (int i = 0; i < 5 && i < flattened.size(); i++) 
        std::cout << flattened[i] << " ";
    std::cout << "\nLast 5: ";
    for (int i = std::max(0, (int)flattened.size() - 5); i < flattened.size(); i++)
        std::cout << flattened[i] << " ";
    std::cout << "\n";

    // Verify we got all nodes
    if (flattened.size() == TOTAL_IPS) {
        std::cout << "Test passed!\n";
    } else {
        std::cout << "Test failed! Expected " << TOTAL_IPS 
                  << " nodes, got " << flattened.size() << "\n";
    }
}

int main() {
    test_big_graph();
    return 0;
}
